tools/fwinfo.py

```diff
@@ -0,0 +1,166 @@
+        if framework_name == stored_framework.name:
+            return stored_framework
+    return None
+
+def get_frameworks():
+    return _framework_infos
+
+def get_framework_names():
+    return [framework.name for framework in get_frameworks()]
+
+def have_framework(framework_name):
+    if get_framework(framework_name):
+        return True
+    return False
+
+def autodiscover_frameworks(repo_root=None):
+    if not repo_root:
+        repo_root=_repo_root
+    if not repo_root:
+        raise Exception("No repo_root provided directly or via init_repo_root()")
+    if _framework_infos:
+        logger.warning( "discover_frameworks() called after frameworks already identified.  Cowardly ignoring.")
+        return
+    frameworks_dir = os.path.join(repo_root, 'frameworks')
+    frameworks = os.listdir(frameworks_dir)
+    for framework in frameworks:
+        add_framework(framework, repo_root)
+
+def shuffle_order():
+    random.shuffle(_framework_infos)
+
+def running_frameworks():
+    return [framework for framework in get_frameworks() if framework.running]
+
+
+class FrameworkTestInfo(object):
+    def __init__(self, framework_name, repo_root):
+        self.name = framework_name
+        self.built = False
+        self.running = False
+        self.cluster = None
+        self.test_success = None # set to True or False later
+        self.stub_universe_url = None # url where cluster can download framework
+        self.output_file = None # in background
+        self.popen = None
+        self.dir = os.path.join(repo_root, 'frameworks', self.name)
+        self.buildscript = os.path.join(self.dir, 'build.sh')
+        # TODO figure out what this trailing slash is for and eliminate
+        self.testdir = os.path.join(self.dir, 'tests') + "/"
+        self.actions = collections.OrderedDict() # succeeded and failed steps land here
+
+    def __repr__(self):
+        # <classname frameworkname>
+        return "<%s %s>" % (self.__class__.__name__, self.name)
+
+    def start_action(self, name):
+        self.actions[name] = {'start': time.time()}
+
+    def finish_action_ok(self, name, **args):
+        action = self.actions[name]
+        action['ok'] = True
+        action['finish'] = time.time()
+        self._complete_action(action, **args)
+
+    def finish_action_fail(self, name, **args):
+        action = self.actions[name]
+        action['ok'] = False
+        action['finish'] = time.time()
+        self._complete_action(action, **args)
+
+    def _complete_action(self, action, **args):
+        action['stdout'] = args.get('stdout')
+        action['stderr'] = args.get('stderr')
+        # failures
+        action['error_message'] = args.get('error_message')
+        action['error_output'] = args.get('error_output')
+
+if __name__ == "__main__":
+    import unittest
+    rel_root_path = os.path.dirname(os.path.dirname(__file__))
+    abs_repo_root = os.path.abspath(rel_root_path)
+
+    class tests(unittest.TestCase):
```

Comment:
I think inline tests for a relatively small module are virtuous.  But I'm not sure I got your point.